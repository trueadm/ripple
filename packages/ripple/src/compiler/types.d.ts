import * as a from '#acorn';

export type DeclarationKind = 'let' | 'const' | 'var' | 'import' | 'function' | 'component' | 'rest_param' | 'param' | 'using' | 'await using';
export type BindingKind = 'normal';

// I think this stuff comes from svelte, should it get removed at some point?
export type LetDirective = unknown;
export type Binding = { kind: BindingKind; [key: string]: unknown; };
