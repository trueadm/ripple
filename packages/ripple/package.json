{
  "name": "ripple",
  "description": "Ripple is an elegant TypeScript UI framework",
  "license": "MIT",
  "author": "Dominic Gannaway",
  "version": "0.2.67",
  "type": "module",
  "module": "src/runtime/index.js",
  "main": "src/runtime/index.js",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/trueadm/ripple.git",
    "directory": "packages/ripple"
  },
  "bugs": {
    "url": "https://github.com/trueadm/ripple/issues"
  },
  "homepage": "https://ripplejs.com",
  "keywords": [
    "ripple",
    "UI",
    "framework"
  ],
  "exports": {
    ".": {
      "types": "./types/index.d.ts",
      "worker": "./src/runtime/index.js",
      "browser": "./src/runtime/index.js",
      "default": "./src/runtime/index.js"
    },
    "./package.json": "./package.json",
    "./compiler": {
      "types": "./types/index.d.ts",
      "require": "./compiler/index.js",
      "default": "./src/compiler/index.js"
    },
    "./validator": {
      "types": "./types/index.d.ts",
      "require": "./validator/index.js",
      "default": "./src/validator/index.js"
    },
    "./internal/client": {
      "default": "./src/runtime/internal/client/index.js"
    },
    "./jsx-runtime": {
      "types": "./src/jsx-runtime.d.ts",
      "import": "./src/jsx-runtime.js",
      "default": "./src/jsx-runtime.js"
    }
  },
  "imports": {
    "#client": "./src/runtime/internal/client/types.d.ts",
    "#client/constants": "./src/internal/client/constants.js"
  },
  "dependencies": {
    "@jridgewell/sourcemap-codec": "^1.5.5",
    "acorn": "^8.15.0",
    "acorn-typescript": "^1.4.13",
    "esrap": "^2.1.0",
    "is-reference": "^3.0.3",
    "magic-string": "^0.30.18",
    "muggle-string": "^0.4.1",
    "zimmerframe": "^1.1.2"
  },
  "devDependencies": {
    "@types/estree": "^1.0.8",
    "@types/node": "^24.3.0",
    "typescript": "catalog:"
  }
}
