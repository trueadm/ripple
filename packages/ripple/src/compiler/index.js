/** @import { Program } from 'estree' */

import { parse as parse_module } from './phases/1-parse/index.js';
import { analyze } from './phases/2-analyze/index.js';
import { transform_client } from './phases/3-transform/client/index.js';
import { transform_server } from './phases/3-transform/server/index.js';
import { convert_source_map_to_mappings } from './phases/3-transform/segments.js';

/**
 * Parse Ripple source code to ESTree AST
 * @param {string} source
 * @returns {Program}
 */
export function parse(source) {
  return parse_module(source);
}

/**
 * Compile Ripple source code to JS/CSS output
 * @param {string} source
 * @param {string} filename
 * @param {{ mode?: 'client' | 'server' }} [options]
 * @returns {object}
 */
export function compile(source, filename, options = {}) {
  const ast = parse_module(source);
  const analysis = analyze(ast, filename);
  const result = options.mode === 'server'
      ? transform_server(filename, source, analysis)
      : transform_client(filename, source, analysis, false);

  return result;
}

/**
 * Compile Ripple source to Volar mappings for editor integration
 * @param {string} source
 * @param {string} filename
 * @returns {object} Volar mappings object
 */
export function compile_to_volar_mappings(source, filename) {
  // Parse and transform
  const ast = parse_module(source);

  // Add unique IDs to import declarations before transformation
  // This allows us to match source imports with generated imports reliably
	// This strategy can potentially be used for other node types in the future
  let gen_id = 0;
  const source_import_map = new Map();
  for (const node of ast.body) {
    if (node.type === 'ImportDeclaration') {
      const start = /** @type {any} */ (node).start;
      const end = /** @type {any} */ (node).end;
      if (start !== undefined && end !== undefined) {
        // Add a unique ID as a string property that will be copied during transformation
        const id = `__volar_import_${gen_id++}__`;
        /** @type {any} */ (node).__volar_id = id;
        source_import_map.set(id, { start, end });
      }
    }
  }

  const analysis = analyze(ast, filename);
  const transformed = transform_client(filename, source, analysis, true);

  // Create volar mappings directly from the AST instead of relying on esrap's sourcemap
  return convert_source_map_to_mappings(transformed.ast, source, transformed.js.code, source_import_map);
}
