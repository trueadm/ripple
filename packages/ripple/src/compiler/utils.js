import { build_assignment_value } from '../utils/ast.js';
import * as b from '../utils/builders.js';
import { get_attribute_event_name, is_delegated, is_event_attribute } from '../utils/events.js';

const regex_return_characters = /\r/g;

const VOID_ELEMENT_NAMES = [
  'area',
  'base',
  'br',
  'col',
  'command',
  'embed',
  'hr',
  'img',
  'input',
  'keygen',
  'link',
  'meta',
  'param',
  'source',
  'track',
  'wbr',
];

/**
 * Returns `true` if `name` is of a void element
 * @param {string} name
 */
export function is_void_element(name) {
  return VOID_ELEMENT_NAMES.includes(name) || name.toLowerCase() === '!doctype';
}

const RESERVED_WORDS = [
  'arguments',
  'await',
  'break',
  'case',
  'catch',
  'class',
  'const',
  'continue',
  'debugger',
  'default',
  'delete',
  'do',
  'else',
  'enum',
  'eval',
  'export',
  'extends',
  'false',
  'finally',
  'for',
  'function',
  'if',
  'implements',
  'import',
  'in',
  'instanceof',
  'interface',
  'let',
  'new',
  'null',
  'package',
  'private',
  'protected',
  'public',
  'return',
  'static',
  'super',
  'switch',
  'this',
  'throw',
  'true',
  'try',
  'typeof',
  'var',
  'void',
  'while',
  'with',
  'yield',
];

export function is_reserved(word) {
  return RESERVED_WORDS.includes(word);
}

/**
 * Attributes that are boolean, i.e. they are present or not present.
 */
const DOM_BOOLEAN_ATTRIBUTES = [
  'allowfullscreen',
  'async',
  'autofocus',
  'autoplay',
  'checked',
  'controls',
  'default',
  'disabled',
  'formnovalidate',
  'hidden',
  'indeterminate',
  'inert',
  'ismap',
  'loop',
  'multiple',
  'muted',
  'nomodule',
  'novalidate',
  'open',
  'playsinline',
  'readonly',
  'required',
  'reversed',
  'seamless',
  'selected',
  'webkitdirectory',
  'defer',
  'disablepictureinpicture',
  'disableremoteplayback',
];

export function is_boolean_attribute(name) {
  return DOM_BOOLEAN_ATTRIBUTES.includes(name);
}

const DOM_PROPERTIES = [
  ...DOM_BOOLEAN_ATTRIBUTES,
  'formNoValidate',
  'isMap',
  'noModule',
  'playsInline',
  'readOnly',
  'value',
  'volume',
  'defaultValue',
  'defaultChecked',
  'srcObject',
  'noValidate',
  'allowFullscreen',
  'disablePictureInPicture',
  'disableRemotePlayback',
];

export function is_dom_property(name) {
  return DOM_PROPERTIES.includes(name);
}

const unhoisted = { hoisted: false };

export function get_delegated_event(event_name, handler, state) {
  // Handle delegated event handlers. Bail out if not a delegated event.
  if (!handler || !is_delegated(event_name)) {
    return null;
  }

  /** @type {FunctionExpression | FunctionDeclaration | ArrowFunctionExpression | null} */
  let target_function = null;
  let binding = null;

  if (handler.type === 'ArrowFunctionExpression' || handler.type === 'FunctionExpression') {
    target_function = handler;
  } else if (handler.type === 'Identifier') {
    binding = state.scope.get(handler.name);

    if (state.analysis.module.scope.references.has(handler.name)) {
      // If a binding with the same name is referenced in the module scope (even if not declared there), bail out
      return unhoisted;
    }

    if (binding != null) {
      for (const { path } of binding.references) {
        const parent = path.at(-1);
        if (parent === undefined) return unhoisted;

        const grandparent = path.at(-2);

        /** @type {AST.RegularElement | null} */
        let element = null;
        /** @type {string | null} */
        let event_name = null;
        if (
          parent.type === 'ExpressionTag' &&
          grandparent?.type === 'Attribute' &&
          is_event_attribute(grandparent)
        ) {
          element = /** @type {AST.RegularElement} */ (path.at(-3));
          const attribute = /** @type {AST.Attribute} */ (grandparent);
          event_name = get_attribute_event_name(attribute.name);
        }

        if (element && event_name) {
          if (
            element.type !== 'Element' ||
            element.metadata.has_spread ||
            !is_delegated(event_name)
          ) {
            return unhoisted;
          }
        } else if (parent.type !== 'FunctionDeclaration' && parent.type !== 'VariableDeclarator') {
          return unhoisted;
        }
      }
    }

    // If the binding is exported, bail out
    if (state.analysis.exports.find((node) => node.name === handler.name)) {
      return unhoisted;
    }

    if (binding !== null && binding.initial !== null && !binding.updated && !binding.is_called) {
      const binding_type = binding.initial.type;

      if (
        binding_type === 'ArrowFunctionExpression' ||
        binding_type === 'FunctionDeclaration' ||
        binding_type === 'FunctionExpression'
      ) {
        target_function = binding.initial;
      }
    }
  }

  // If we can't find a function, or the function has multiple parameters, bail out
  if (target_function == null || target_function.params.length > 1) {
    return unhoisted;
  }

  const visited_references = new Set();
  const scope = target_function.metadata.scope;
  for (const [reference] of scope.references) {
    // Bail out if the arguments keyword is used or $host is referenced
    if (reference === 'arguments') return unhoisted;

    const binding = scope.get(reference);
    const local_binding = state.scope.get(reference);

    // If we are referencing a binding that is shadowed in another scope then bail out.
    if (local_binding !== null && binding !== null && local_binding.node !== binding.node) {
      return unhoisted;
    }

    if (
      binding !== null &&
      // Bail out if the the binding is a rest param
      (binding.declaration_kind === 'rest_param' || // or any normal not reactive bindings that are mutated.
        // Bail out if we reference anything from the EachBlock (for now) that mutates in non-runes mode,
        (binding.kind === 'normal' && binding.updated))
    ) {
      return unhoisted;
    }
    visited_references.add(reference);
  }

  return { hoisted: true, function: target_function };
}

function get_hoisted_params(node, context) {
  const scope = context.state.scope;

  /** @type {Identifier[]} */
  const params = [];

  /**
   * We only want to push if it's not already present to avoid name clashing
   * @param {Identifier} id
   */
  function push_unique(id) {
    if (!params.find((param) => param.name === id.name)) {
      params.push(id);
    }
  }

  for (const [reference] of scope.references) {
    let binding = scope.get(reference);

    if (binding !== null && !scope.declarations.has(reference) && binding.initial !== node) {
      if (binding.kind === 'prop') {
        push_unique(b.id('__props'));
      } else if (binding.kind === 'prop_fallback') {
        push_unique(b.id(binding.node.name));
      } else if (
        // imports don't need to be hoisted
        binding.declaration_kind !== 'import'
      ) {
        // create a copy to remove start/end tags which would mess up source maps
        push_unique(b.id(binding.node.name));
      }
    }
  }
  return params;
}

export function build_hoisted_params(node, context) {
  const hoisted_params = get_hoisted_params(node, context);
  node.metadata.hoisted_params = hoisted_params;

  /** @type {Pattern[]} */
  const params = [];

  if (node.params.length === 0) {
    if (hoisted_params.length > 0) {
      // For the event object
      params.push(b.id(context.state.scope.generate('_')));
    }
  } else {
    for (const param of node.params) {
      params.push(/** @type {Pattern} */ (context.visit(param)));
    }
  }

  params.push(...hoisted_params, b.id('__block'));
  return params;
}

export function is_top_level_await(context) {
  if (!is_inside_component(context)) {
    return false;
  }

  for (let i = context.path.length - 1; i >= 0; i -= 1) {
    const context_node = context.path[i];
    const type = context_node.type;

    if (type === 'Component') {
      return true;
    }

    if (
      type === 'FunctionExpression' ||
      type === 'ArrowFunctionExpression' ||
      type === 'FunctionDeclaration'
    ) {
      return false;
    }
  }
  return true;
}

export function is_inside_component(context, includes_functions = false) {
  for (let i = context.path.length - 1; i >= 0; i -= 1) {
    const context_node = context.path[i];
    const type = context_node.type;

    if (
      !includes_functions &&
      (type === 'FunctionExpression' ||
        type === 'ArrowFunctionExpression' ||
        type === 'FunctionDeclaration')
    ) {
      return false;
    }
    if (type === 'Component') {
      return true;
    }
  }
  return false;
}

export function is_component_level_function(context) {
  for (let i = context.path.length - 1; i >= 0; i -= 1) {
    const context_node = context.path[i];
    const type = context_node.type;

    if (type === 'BlockStatement' && context_node.body.find((n) => n.type === 'Component')) {
      return true;
    }

    if (
      type === 'FunctionExpression' ||
      type === 'ArrowFunctionExpression' ||
      type === 'FunctionDeclaration'
    ) {
      return false;
    }
  }
  return true;
}

export function is_inside_call_expression(context) {
  for (let i = context.path.length - 1; i >= 0; i -= 1) {
    const context_node = context.path[i];
    const type = context_node.type;

    if (
      type === 'FunctionExpression' ||
      type === 'ArrowFunctionExpression' ||
      type === 'FunctionDeclaration'
    ) {
      return false;
    }
    if (type === 'CallExpression') {
      return true;
    }
  }
  return false;
}

export function is_value_static(node) {
  if (node.type === 'Literal') {
    return true;
  }
  if (node.type === 'ArrayExpression') {
    return true;
  }
  if (node.type === 'NewExpression') {
    if (node.callee.type === 'Identifier' && node.callee.name === 'Array') {
      return true;
    }
    return false;
  }

  return false;
}

export function is_ripple_import(callee, context) {
  if (callee.type === 'Identifier') {
    const binding = context.state.scope.get(callee.name);

    return (
      binding?.declaration_kind === 'import' &&
      binding.initial.source.type === 'Literal' &&
      binding.initial.source.value === 'ripple'
    );
  } else if (
    callee.type === 'MemberExpression' &&
    callee.object.type === 'Identifier' &&
    !callee.computed
  ) {
    const binding = context.state.scope.get(callee.object.name);

    return (
      binding?.declaration_kind === 'import' &&
      binding.initial.source.type === 'Literal' &&
      binding.initial.source.value === 'ripple'
    );
  }

  return false;
}

export function is_declared_function_within_component(node, context) {
  const component = context.path.find((n) => n.type === 'Component');

  if (node.type === 'Identifier' && component) {
    const binding = context.state.scope.get(node.name);
    const component_scope = context.state.scopes.get(component);

    if (binding !== null && component_scope !== null) {
      if (
        binding.declaration_kind !== 'function' &&
        binding.initial?.type !== 'FunctionDeclaration' &&
        binding.initial?.type !== 'ArrowFunctionExpression' &&
        binding.initial?.type !== 'FunctionExpression'
      ) {
        return false;
      }
      let scope = binding.scope;

      while (scope !== null) {
        if (scope === component_scope) {
          return true;
        }
        scope = scope.parent;
      }
    }
  }

  return false;
}

function is_non_coercive_operator(operator) {
  return ['=', '||=', '&&=', '??='].includes(operator);
}

export function visit_assignment_expression(node, context, build_assignment) {
  if (
    node.left.type === 'ArrayPattern' ||
    node.left.type === 'ObjectPattern' ||
    node.left.type === 'RestElement'
  ) {
    const value = /** @type {Expression} */ (context.visit(node.right));
    const should_cache = value.type !== 'Identifier';
    const rhs = should_cache ? b.id('$$value') : value;

    let changed = false;

    const assignments = extract_paths(node.left).map((path) => {
      const value = path.expression?.(rhs);

      let assignment = build_assignment('=', path.node, value, context);
      if (assignment !== null) changed = true;

      return (
        assignment ??
        b.assignment(
          '=',
          /** @type {Pattern} */ (context.visit(path.node)),
          /** @type {Expression} */ (context.visit(value)),
        )
      );
    });

    if (!changed) {
      // No change to output -> nothing to transform -> we can keep the original assignment
      return null;
    }

    const is_standalone = /** @type {Node} */ (context.path.at(-1)).type.endsWith('Statement');
    const sequence = b.sequence(assignments);

    if (!is_standalone) {
      // this is part of an expression, we need the sequence to end with the value
      sequence.expressions.push(rhs);
    }

    if (should_cache) {
      // the right hand side is a complex expression, wrap in an IIFE to cache it
      const iife = b.arrow([rhs], sequence);

      const iife_is_async =
        is_expression_async(value) ||
        assignments.some((assignment) => is_expression_async(assignment));

      return iife_is_async ? b.await(b.call(b.async(iife), value)) : b.call(iife, value);
    }

    return sequence;
  }

  if (node.left.type !== 'Identifier' && node.left.type !== 'MemberExpression') {
    throw new Error(`Unexpected assignment type ${node.left.type}`);
  }

  const transformed = build_assignment(node.operator, node.left, node.right, context);

  if (transformed === node.left) {
    return node;
  }

  return transformed;
}

export function build_assignment(operator, left, right, context) {
  let object = left;

  while (object.type === 'MemberExpression') {
    // @ts-expect-error
    object = object.object;
  }

  if (object.type !== 'Identifier') {
    return null;
  }

  const binding = context.state.scope.get(object.name);
  if (!binding) return null;

  const transform = binding.transform;

  // reassignment
  if (object === left || (left.type === 'MemberExpression' && left.computed && operator === '=')) {
    const assign_fn = transform?.assign || transform?.assign_tracked;
    if (assign_fn) {
      let value = /** @type {Expression} */ (
        context.visit(build_assignment_value(operator, left, right))
      );

      return assign_fn(
        object,
        value,
        left.type === 'MemberExpression' && left.computed
          ? context.visit(left.property)
          : undefined,
      );
    }
  }

  // mutation
  if (transform?.mutate) {
    return transform.mutate(
      object,
      b.assignment(
        operator,
        /** @type {Pattern} */ (context.visit(left)),
        /** @type {Expression} */ (context.visit(right)),
      ),
    );
  }

  return null;
}

const ATTR_REGEX = /[&"<]/g;
const CONTENT_REGEX = /[&<]/g;

export function escape_html(value, is_attr = false) {
  const str = String(value ?? '');

  const pattern = is_attr ? ATTR_REGEX : CONTENT_REGEX;
  pattern.lastIndex = 0;

  let escaped = '';
  let last = 0;

  while (pattern.test(str)) {
    const i = pattern.lastIndex - 1;
    const ch = str[i];
    escaped += str.substring(last, i) + (ch === '&' ? '&amp;' : ch === '"' ? '&quot;' : '&lt;');
    last = i + 1;
  }

  return escaped + str.substring(last);
}

export function hash(str) {
  str = str.replace(regex_return_characters, '');
  let hash = 5381;
  let i = str.length;

  while (i--) hash = ((hash << 5) - hash) ^ str.charCodeAt(i);
  return (hash >>> 0).toString(36);
}

export function is_element_dom_element(node) {
  return (
    node.id.type === 'Identifier' &&
    node.id.name[0].toLowerCase() === node.id.name[0] &&
    node.id.name !== 'children' &&
    !node.id.tracked
  );
}
