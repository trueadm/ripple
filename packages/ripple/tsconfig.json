{
	"compilerOptions": {
		"module": "esnext",
		"lib": ["esnext", "dom", "dom.iterable"],
		"target": "esnext",
		"noEmit": true,
		"moduleResolution": "bundler",
		"resolveJsonModule": true,
		"noEmitOnError": true,
		"noErrorTruncation": true,
		"allowSyntheticDefaultImports": true,
		"verbatimModuleSyntax": true,
		"types": ["node"],
		"jsx": "preserve",
		"jsxImportSource": "ripple",
		"strict": true,
		"allowJs": true,
		"checkJs": true,
		"paths": {
			"ripple": ["./types/index.d.ts"],
		}
	},
	"include": [
		"./*.js",
		"./src/",
		"./tests/*/test.ripple",
	]
}
