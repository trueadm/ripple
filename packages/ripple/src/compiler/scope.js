/** @import { ClassDeclaration, Expression, FunctionDeclaration, FunctionExpression, Identifier, ImportDeclaration, MemberExpression, LogicalOperator, Node, Pattern, UnaryOperator, VariableDeclarator, Super } from 'estree' */
/** @import { DeclarationKind, BindingKind, Component, Element } from '#compiler' */
/** @import { RippleNode } from '#compiler' */

import is_reference from 'is-reference';
import { extract_identifiers, object, unwrap_pattern } from '../utils/ast.js';
import { walk } from 'zimmerframe';
import { is_reserved } from './utils.js';
import * as b from '../utils/builders.js';

export class Binding {
	/** @type {Scope} */
	scope;

	/** @type {Identifier} */
	node;

	/** @type {BindingKind} */
	kind;

	/** @type {DeclarationKind} */
	declaration_kind;

	/**
	 * What the value was initialized with
	 * @type {null | Expression | FunctionDeclaration | ClassDeclaration | ImportDeclaration}
	 */
	initial = null;

	/** @type {Array<{ node: Identifier; path: RippleNode[] }>} */
	references = [];

	mutated = false;
	reassigned = false;
	is_called = false;
	metadata = null;

	/**
	 * @param {Scope} scope
	 * @param {Identifier} node
	 * @param {BindingKind} kind
	 * @param {DeclarationKind} declaration_kind
	 * @param {Binding['initial']} initial
	 */
	constructor(scope, node, kind, declaration_kind, initial) {
		this.scope = scope;
		this.node = node;
		this.initial = initial;
		this.kind = kind;
		this.declaration_kind = declaration_kind;
	}

	get updated() {
		return this.mutated || this.reassigned;
	}
}

/**
 * Create scopes for an AST
 * @param {Node} ast - The AST to create scopes for
 * @param {ScopeRoot} root - Root scope manager
 * @param {Scope | null} parent - Parent scope
 * @returns {{ scope: Scope, scopes: Map<RippleNode, Scope> }} Scope information
 */
export function create_scopes(ast, root, parent) {
	/** @typedef {{ scope: Scope }} State */

	/** @type {Map<RippleNode, Scope>} */
	const scopes = new Map();
	const scope = new Scope(root, parent, false);
	scopes.set(ast, scope);

	/** @type {State} */
	const state = { scope };
	/** @type {Array<[Scope, { node: Identifier, path: RippleNode[] }]>} */
	const references = [];
	/** @type {Array<[Scope, Pattern | MemberExpression]>} */
	const updates = [];

	/**
	 * Add parameters to scope
	 * @param {Scope} scope - The scope to add parameters to
	 * @param {Pattern[]} params - Parameter nodes
	 */
	function add_params(scope, params) {
		for (const param of params) {
			for (const node of extract_identifiers(param)) {
				scope.declare(node, 'normal', param.type === 'RestElement' ? 'rest_param' : 'param');
			}
		}
	}

	/**
	 * Create a block scope
	 * @param {any} node - AST node
	 * @param {{ state: any, next: Function }} context - Visitor context
	 */
	const create_block_scope = (node, { state, next }) => {
		const scope = state.scope.child(true);
		scopes.set(node, scope);

		if (node.type === 'ForOfStatement') {
			if (node.index) {
				state.scope.declare(node.index, 'normal', 'let');
			}
		}

		next({ scope });
	};

	walk(/** @type {RippleNode} */ (ast), state, {
		// references
		Identifier(node, { path, state }) {
			const parent = path.at(-1);
			if (
				parent &&
				is_reference(node, /** @type {Node} */ (parent)) &&
				// TSTypeAnnotation, TSInterfaceDeclaration etc - these are normally already filtered out,
				// but for the migration they aren't, so we need to filter them out here
				// TODO -> once migration script is gone we can remove this check
				!parent.type.startsWith('TS')
			) {
				references.push([state.scope, { node, path: path.slice() }]);
			}
		},

		AssignmentExpression(node, { state, next }) {
			updates.push([state.scope, node.left]);
			next();
		},

		UpdateExpression(node, { state, next }) {
			updates.push([state.scope, /** @type {Identifier | MemberExpression} */ (node.argument)]);
			next();
		},

		ImportDeclaration(node, { state }) {
			for (const specifier of node.specifiers) {
				state.scope.declare(specifier.local, 'normal', 'import', node);
			}
		},

		/**
		 * @param {Component} node
		 * @param {Object} context
		 * @param {any} context.state
		 * @param {Function} context.next
		 */
		Component(node, { state, next }) {
			const scope = state.scope.child();
			scopes.set(node, scope);

			scope.declare(node.id, 'normal', 'component');

			add_params(scope, node.params);
			next({ scope });
		},

		/**
		 * @param {Element} node
		 * @param {Object} context
		 * @param {any} context.state
		 * @param {Function} context.next
		 */
		Element(node, { state, next }) {
			const scope = state.scope.child();
			scopes.set(node, scope);

			next({ scope });
		},

		ServerBlock(node, { state, next }) {
			const scope = state.scope.child();
			scope.server_block = true;
			scopes.set(node, scope);

			next({ scope });
		},

		FunctionExpression(node, { state, next }) {
			const scope = state.scope.child();
			scopes.set(node, scope);

			if (node.id) scope.declare(node.id, 'normal', 'function');

			add_params(scope, node.params);
			next({ scope });
		},

		FunctionDeclaration(node, { state, next }) {
			if (node.id) state.scope.declare(node.id, 'normal', 'function', node);

			const scope = state.scope.child();
			scopes.set(node, scope);

			add_params(scope, node.params);
			next({ scope });
		},

		ArrowFunctionExpression(node, { state, next }) {
			const scope = state.scope.child();
			scopes.set(node, scope);

			add_params(scope, node.params);
			next({ scope });
		},

		ForStatement: create_block_scope,
		ForInStatement: create_block_scope,
		ForOfStatement: create_block_scope,
		SwitchStatement: create_block_scope,
		BlockStatement(node, context) {
			const parent = context.path.at(-1);
			if (
				parent?.type === 'FunctionDeclaration' ||
				parent?.type === 'FunctionExpression' ||
				parent?.type === 'ArrowFunctionExpression'
			) {
				// We already created a new scope for the function
				context.next();
			} else {
				create_block_scope(node, context);
			}
		},

		ClassDeclaration(node, { state, next }) {
			if (node.id) state.scope.declare(node.id, 'normal', 'let', node);
			next();
		},

		VariableDeclaration(node, { state, path, next }) {
			for (const declarator of node.declarations) {
				/** @type {Binding[]} */
				const bindings = [];

				state.scope.declarators.set(declarator, bindings);

				for (const id of extract_identifiers(declarator.id)) {
					const binding = state.scope.declare(id, 'normal', node.kind, declarator.init);
					bindings.push(binding);
				}
			}

			next();
		},

		CatchClause(node, { state, next }) {
			if (node.param) {
				const scope = state.scope.child(true);
				scopes.set(node, scope);

				for (const id of extract_identifiers(node.param)) {
					scope.declare(id, 'normal', 'let');
				}

				next({ scope });
			} else {
				next();
			}
		},
	});

	for (const [scope, { node, path }] of references) {
		scope.reference(node, path);
	}

	for (const [scope, node] of updates) {
		for (const expression of unwrap_pattern(node)) {
			const left = object(expression);
			const binding = left && scope.get(left.name);

			if (binding !== null && left !== binding.node) {
				binding.updated = true;

				if (left === expression) {
					binding.reassigned = true;
				} else {
					binding.mutated = true;
				}
			}
		}
	}

	return {
		scope,
		scopes,
	};
}

export class Scope {
	/** @type {ScopeRoot} */
	root;

	/**
	 * The immediate parent scope
	 * @type {Scope | null}
	 */
	parent;

	/**
	 * Whether or not `var` declarations are contained by this scope
	 * @type {boolean}
	 */
	#porous;

	/**
	 * A map of every identifier declared by this scope, and all the
	 * identifiers that reference it
	 * @type {Map<string, Binding>}
	 */
	declarations = new Map();

	/**
	 * A map of declarators to the bindings they declare
	 * @type {Map<VariableDeclarator, Binding[]>}
	 */
	declarators = new Map();

	/**
	 * A set of all the names referenced with this scope
	 * â€” useful for generating unique names
	 * @type {Map<string, { node: Identifier; path: RippleNode[] }[]>}
	 */
	references = new Map();

	/**
	 * The scope depth allows us to determine if a state variable is referenced in its own scope,
	 * which is usually an error. Block statements do not increase this value
	 */
	function_depth = 0;

	/**
	 * If tracing of reactive dependencies is enabled for this scope
	 * @type {null | Expression}
	 */
	tracing = null;

	/**
	 * Is this scope a top-level server block scope
	 * @type {boolean}
	 */
	server_block = false;

	/**
	 *
	 * @param {ScopeRoot} root
	 * @param {Scope | null} parent
	 * @param {boolean} porous
	 */
	constructor(root, parent, porous) {
		this.root = root;
		this.parent = parent;
		this.#porous = porous;
		this.function_depth = parent ? parent.function_depth + (porous ? 0 : 1) : 0;
	}

	/**
	 * @param {Identifier} node
	 * @param {Binding['kind']} kind
	 * @param {DeclarationKind} declaration_kind
	 * @param {null | Expression | FunctionDeclaration | ClassDeclaration | ImportDeclaration} initial
	 * @returns {Binding}
	 */
	declare(node, kind, declaration_kind, initial = null) {
		if (this.parent) {
			if (declaration_kind === 'var' && this.#porous) {
				return this.parent.declare(node, kind, declaration_kind);
			}

			if (declaration_kind === 'import' && !this.parent.server_block) {
				return this.parent.declare(node, kind, declaration_kind, initial);
			}
		}

		if (node.name === '_$_') {
			throw new Error('Cannot declare a variable named "_$_" as it is a reserved identifier');
		}

		if (this.declarations.has(node.name)) {
			throw new Error(`'${node.name}' has already been declared in the current scope`);
		}

		/** @type {Binding} */
		const binding = {
			node,
			references: [],
			initial,
			reassigned: false,
			mutated: false,
			updated: false,
			scope: this,
			kind,
			declaration_kind,
			is_called: false,
			metadata: null,
		};

		this.declarations.set(node.name, binding);
		this.root.conflicts.add(node.name);
		return binding;
	}

	child(porous = false) {
		return new Scope(this.root, this, porous);
	}

	/**
	 * @param {string} preferred_name
	 * @returns {string}
	 */
	generate(preferred_name) {
		if (this.#porous) {
			return /** @type {Scope} */ (this.parent).generate(preferred_name);
		}

		preferred_name = preferred_name.replace(/[^a-zA-Z0-9_$]/g, '_').replace(/^[0-9]/, '_');
		let name = preferred_name;
		let n = 1;

		while (
			this.references.has(name) ||
			this.declarations.has(name) ||
			this.root.conflicts.has(name) ||
			is_reserved(name)
		) {
			name = `${preferred_name}_${n++}`;
		}

		this.references.set(name, []);
		this.root.conflicts.add(name);
		return name;
	}

	/**
	 * @param {string} name
	 * @returns {Binding | null}
	 */
	get(name) {
		return this.declarations.get(name) ?? this.parent?.get(name) ?? null;
	}

	/**
	 * @param {VariableDeclarator} node
	 * @returns {Binding[]}
	 */
	get_bindings(node) {
		const bindings = this.declarators.get(node);
		if (!bindings) {
			throw new Error('No binding found for declarator');
		}
		return bindings;
	}

	/**
	 * @param {string} name
	 * @returns {Scope | null}
	 */
	owner(name) {
		return this.declarations.has(name) ? this : this.parent && this.parent.owner(name);
	}

	/**
	 * @param {Identifier} node
	 * @param {RippleNode[]} path
	 */
	reference(node, path) {
		path = [...path]; // ensure that mutations to path afterwards don't affect this reference
		let references = this.references.get(node.name);

		if (!references) this.references.set(node.name, (references = []));

		references.push({ node, path });

		const binding = this.declarations.get(node.name);
		if (binding) {
			binding.references.push({ node, path });
		} else if (this.parent) {
			this.parent.reference(node, path);
		} else {
			// no binding was found, and this is the top level scope,
			// which means this is a global
			this.root.conflicts.add(node.name);
		}
	}
}

export class ScopeRoot {
	/** @type {Set<string>} */
	conflicts = new Set();

	/**
	 * @param {string} preferred_name
	 */
	unique(preferred_name) {
		preferred_name = preferred_name.replace(/[^a-zA-Z0-9_$]/g, '_');
		let final_name = preferred_name;
		let n = 1;

		while (this.conflicts.has(final_name)) {
			final_name = `${preferred_name}_${n++}`;
		}

		this.conflicts.add(final_name);
		const id = b.id(final_name);
		return id;
	}
}
