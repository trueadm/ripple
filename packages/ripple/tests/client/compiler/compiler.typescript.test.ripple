import { compile } from 'ripple/compiler';

describe('compiler > typescript', () => {
	it('compiles TSInstantiationExpression', () => {
		const source =
`function makeBox<T>(value: T) {
	return { value };
}
const makeStringBox = makeBox<string>;
const stringBox = makeStringBox('abc');
const ErrorMap = Map<string, Error>;
const errorMap = new ErrorMap();`;

		const result = compile(source, 'test.ripple', { mode: 'client' });

		expect(result.js.code).toMatchSnapshot();
	});

	it('removes type assertions from function parameters and leaves default values', () => {
		const source = `
function getString(e: string = 'test') {
	return e;
}`;

		const result = compile(source, 'test.ripple', { mode: 'client' });

		expect(result.js.code).toMatchSnapshot();
	});
});
