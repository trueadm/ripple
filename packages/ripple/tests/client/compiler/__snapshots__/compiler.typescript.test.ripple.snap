// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler > typescript > compiles TSInstantiationExpression 1`] = `
"import * as _$_ from 'ripple/internal/client';

function makeBox(value) {
	return { value };
}

const makeStringBox = (makeBox);
const stringBox = makeStringBox('abc');
const ErrorMap = (Map);
const errorMap = new ErrorMap();"
`;

exports[`compiler > typescript > removes type assertions from function parameters and leaves default values 1`] = `
"import * as _$_ from 'ripple/internal/client';

function getString(e = 'test') {
	return e;
}"
`;
